File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N, M, R;
    static int[][] arr;
    // 아래, 오른쪽, 위, 왼쪽
    static int[][] dir = {{0, 1}, {1, 0}, {0, -1}, {-1,0}};

    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        arr = new int[N][M];

        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<M; j++){
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        for(int i=0; i<R; i++){
            rotate();
        }

        for(int a=0; a<N; a++){
            for(int b=0; b<M; b++){
                System.out.print(arr[a][b]+" ");
            }
            System.out.println();
        }

    }
    private static void rotate(){
        int rotateNum = Math.min(N, M);

        for(int i=0; i<rotateNum/2; i++){
            int r = i; int c=i;
            int temp = arr[r][c];
            int idx = 0;
            while(idx<4){
                int nr = r + dir[idx][0];
                int nc = c + dir[idx][1];
                if(nr<N-i && nc<M-i && nr>=i && nc>=i){
                    arr[r][c] = arr[nr][nc];
                    r = nr; c = nc;
                }else {
                    idx++;
                }

            }
            arr[i+1][i] = temp;
        }
    }
}

File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.kt
import java.util.*
import java.io.*

fun main()=with(BufferedReader(InputStreamReader(System.`in`))){
    var st = StringTokenizer(readLine())
    val N = Integer.parseInt(st.nextToken())
    val M = Integer.parseInt(st.nextToken())
    val R = Integer.parseInt(st.nextToken())

    var arr = Array(N){Array(M){0}}
    for(i in 0 until N){
        st = StringTokenizer(readLine())
        for(j in 0 until M){
            arr[i][j] = Integer.parseInt(st.nextToken())
        }
    }
    var cnt = Math.min(N, M)
    cnt /= 2
    for(r in 0 until R){
        var n = N;
        var m = M;
        for(c in 0 until cnt){
            var temp = arr[c][c]
            // <왼
            for(i in 1+c until m){
                arr[c][i-1] = arr[c][i]
            }
            // ^위
            for(i in 1+c until n){
                arr[i-1][m-1] = arr[i][m-1]
            }
            // >오
            for(i in m-1 downTo 1+c){
                arr[n-1][i] = arr[n-1][i-1]
            }

            // V아
            for(i in n-1 downTo 1+c){
                arr[i][c] = arr[i-1][c]
            }
            arr[c+1][c] = temp
            m--
            n--
        }

    }
    var sb = StringBuilder()
    for(i in 0 until N){
        for(j in 0 until M){
            sb.append(arr[i][j]).append(' ');
        }
        sb.append('\n')
    }

    print(sb)
}